src/main/java/teammates/common/datatransfer/FeedbackSessionResultsBundle.java

```diff
@@ -76,645 +76,211 @@
-
-            String giverTeam1 = getTeamNameForEmail(o1.giver).isEmpty() ? getNameForEmail(o1.giver)
-                                                                        : getTeamNameForEmail(o1.giver);
-            String giverTeam2 = getTeamNameForEmail(o2.giver).isEmpty() ? getNameForEmail(o2.giver)
-                                                                        : getTeamNameForEmail(o2.giver);
-            int order = compareByNames(giverTeam1, giverTeam2, isGiverVisible1, isGiverVisible2);
-            if (order != 0) {
-                return order;
-            }
-
-            order = compareByQuestionNumber(o1, o2);
-            if (order != 0) {
-                return order;
-            }
-
-            String giverName1 = emailNameTable.get(o1.giver);
-            String giverName2 = emailNameTable.get(o2.giver);
-            order = compareByNames(giverName1, giverName2, isGiverVisible1, isGiverVisible2);
-            if (order != 0) {
-                return order;
-            }
-
-            boolean isRecipientVisible1 = isRecipientVisible(o1);
-            boolean isRecipientVisible2 = isRecipientVisible(o2);
-
-            String receiverTeam1 = getTeamNameForEmail(o1.recipient).isEmpty() ? getNameForEmail(o1.recipient)
-                                                                               : getTeamNameForEmail(o1.recipient);
-            String receiverTeam2 = getTeamNameForEmail(o2.recipient).isEmpty() ? getNameForEmail(o2.recipient)
-                                                                               : getTeamNameForEmail(o2.recipient);
-            order = compareByNames(receiverTeam1, receiverTeam2, isRecipientVisible1, isRecipientVisible2);
-            if (order != 0) {
-                return order;
-            }
-
-            String recipientName1 = emailNameTable.get(o1.recipient);
-            String recipientName2 = emailNameTable.get(o2.recipient);
-            order = compareByNames(recipientName1, recipientName2, isRecipientVisible1, isRecipientVisible2);
-
-            if (order != 0) {
-                return order;
-            }
-
-            order = compareByResponseString(o1, o2);
-            if (order != 0) {
-                return order;
-            }
-
-            return o1.getId().compareTo(o2.getId());
-        }
-    };
+    private Comparator<FeedbackResponseAttributes> compareByTeamQuestionGiverTeamRecipient = compareByNamesForGiver
+            .thenComparing(compareByQuestionNumber)
+            .thenComparing(compareByGiverInEmailNameTable)
+            .thenComparing(compareByNamesForRecipient)
+            .thenComparing(compareByRecipientInEmailNameTable)
+            .thenComparing(compareByResponseString)
+            .thenComparing(compareByFeedbackResponseAttributeId);
 
     // Sorts by recipientName > recipientEmail > giverName > giverEmail
     private Comparator<FeedbackResponseAttributes> compareByRecipientNameEmailGiverNameEmail =
```

Comment:
The contents of this comparator can benefit from reuse of existing comparators as well, similar to all the comparators before this.

Reply:
I wanted to do the same, but I noticed that this comparator had a bit different condition while trying to compare the names of giver and recipient. Also, the comparison by email is used only in this comparator and hence I didn't find the benefit of extracting it for using it just once. 

E.g. 
- In the extracted comparator of `compareByGiverName` the following code is used: 
```
compareByNames(emailNameTable.get(fra1.giver), emailNameTable.get(fra2.giver),
                            isGiverVisible(fra1), isGiverVisible(fra2)
```

- However, in this particular comparator, the logic used is a bit different:
```
int giverNameCompareResult = compareByNames(getNameForEmail(o1.giver),
                        getNameForEmail(o2.giver),
                        isGiverVisible1, isGiverVisible2);
```

The following is how `getNameFormEmail()` is implemented which is I guess has different behaviour from `emailNameTable.get(o1.giver)`

```
public String getNameForEmail(String email) {
        String name = emailNameTable.get(email);
        if (name == null || name.equals(Const.USER_IS_MISSING)) {
            return Const.USER_UNKNOWN_TEXT;
        } else if (name.equals(Const.USER_IS_NOBODY)) {
            return Const.USER_NOBODY_TEXT;
        } else if (name.equals(Const.USER_IS_TEAM)) {
            return getTeamNameForEmail(email);
        } else {
            return name;
        }
    }
```

Due to this, I avoided simplifying it, however, it can be done if you want it to by extracting them as new comparators.

Reply:
@tshradheya I know they're not the same and I don't ask you to make them the same.

> I didn't find the benefit of extracting it for using it just once

SLAP principle comes in here. Will you be happier if this comparator reads so smoothly like the others, or keeping it like this and additionally knowing full well that there are repetitions (`compareByResponseString`, for instance) merely to satisfy "this block is only used once and should not be extracted out"?